### PR TITLE
Payment Term Discounts

### PR DESCRIPTION
**Pull request by agritheory**
Pull-Request
- [X ] Have you followed the guidelines in our Contributing document?
- [X ] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
- [ X] Have you lint your code locally prior to submission?
- [x] Have you successfully run tests with your changes locally?
- [X ] Does your commit message have an explanation for your changes and why you'd like us to include them?
- [x] Docs have been added / updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Did you modify the existing test cases? If yes, why?

---

What type of a PR is this? 
- [X] New Feature Submissions, do not merge, for code review and feedback
--- 

- Motivation and Context (What existing problem does the pull request solve):
> It is common practice in the US and other countries for a supplier to offer a payment discount if the remittance is within an accelerated period. The most common example I am aware of is "2/10 Net 30", where the buyer is given a 2% discount. While that may sound trivial, the businesses that are organized enough to take advantage of it are getting an annualized discount of 20.41% in the 2/10 Net 30 example.

from #14472 

- Related Issues: 
https://github.com/frappe/erpnext/issues/4353
https://github.com/frappe/erpnext/issues/14472

- Screenshots (if applicable, remember, a picture tells a thousand words): 
Screen shots and description of the changes are in #14472 above.

From that proposal issue, a few changes have been made:
-[X] JS code has been broken up a bit more (though I am certainly open to making it a more readable, it's not awesome, as-is)
-[x] Trivial linting of legacy code
-[ ] I think the set invoices methods need a better, more integrated design, I'd like some feedback on this specifically, because I think it means [changing the accounts controller ](https://github.com/frappe/erpnext/blob/7048925016c080d736ed81105842c36da8f840d4/erpnext/controllers/accounts_controller.py#L942)and gets into a scope-of-changes- creep issue

Status:
This is not a final version of this code, I need to incorporate these changes into the tests and get some feedback on the final step: setting invoices as paid.

### CODE REVIEW

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 1]
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
...
+def cancel_discounted_invoices(transactions):
+	for item in transactions:
+		frappe.db.set_value(item["reference_doctype"], item["reference_document"], "outstanding_amount", 0)
+		frappe.db.set_value(item["reference_doctype"], item["reference_document"], "Status", "Paid")
+		frappe.db.set_value(item["reference_doctype"], item["reference_document"], "discount_amount", item["amount"])
+	frappe.db.commit()
+
+
+@frappe.whitelist()
+def calc_discount(discount_double_check):
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: https://github.com/frappe/erpnext/wiki/Coding-Standards#break-functions--methods-more-than-10-lines-of-code

**Review by rmehta — DISMISSED** [review]
https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

File: erpnext/accounts/doctype/payment_entry/payment_entry.py (hunk 1 above)
Comment (agritheory) [reply erpnext/accounts/doctype/payment_entry/payment_entry.py]: @rmehta This 10 lines guideline: both JS and py? The link doesn't specify if the style guide is the same for both or not.

Comment (rmehta) [reply erpnext/accounts/doctype/payment_entry/payment_entry.py]: Obviously both

File: erpnext/setup/doctype/company/company.json [hunk 2]
```diff
@@ -1,2888 +1,2987 @@
 {
- "allow_copy": 0,
```
Comment (vjFaLk) [inline erpnext/setup/doctype/company/company.json]: Fix formatting

File: erpnext/agriculture/doctype/crop_cycle/crop_cycle.js [hunk 3]
```diff
@@ -1,48 +0,0 @@
-// Copyright (c) 2017, Frappe Technologies Pvt. Ltd. and contributors
```
Comment (vjFaLk) [inline erpnext/agriculture/doctype/crop_cycle/crop_cycle.js]: Accidentally deleted this file

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 4]
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
...
-	return paid_amount[0][0] if paid_amount else 0
\ No newline at end of file
+	return paid_amount[0][0] if paid_amount else 0
+
+
+@frappe.whitelist
+def set_paid_invoices(transactions):
+	frappe.throw("set INvoices")
+	for item in transactions:
+		frappe.db.set_value(item["reference_doctype"], item["reference_document"], "outstanding_amount", 0)
```
Comment (vjFaLk) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: I think there's a way to db_set multiple values at once. Or just do a `get_doc`

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 5]
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
...
 	""".format(dr_or_cr=dr_or_cr), (dt, dn, party_type, party, account, due_date))
 
-	return paid_amount[0][0] if paid_amount else 0
\ No newline at end of file
+	return paid_amount[0][0] if paid_amount else 0
+
+
+@frappe.whitelist
+def set_paid_invoices(transactions):
+	frappe.throw("set INvoices")
```
Comment (vjFaLk) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: Bruh 💩

File: erpnext/accounts/doctype/payment_terms_template/payment_terms_template.js [hunk 6]
```diff
@@ -8,5 +8,22 @@ frappe.ui.form.on('Payment Terms Template', {
...
+		frm.add_fetch("payment_term", "discount_eligible_percent", "discount_eligible_percent");
+	}, validate: function(frm){
+		for(let i in frm.doc.terms){
+			if(frm.doc.terms[i]["discount_eligible_days"] > frm.doc.terms[i]["credit_days"]){
+				frappe.throw("Discount Days must be less than Credit Days."); // REVIEW: should this be a server method?
+			}
+			if(frm.doc.terms[i]["discount_percent"] < 0){
+				frappe.throw("Please enter the discount as a positive number.");
+			}
+			if(frm.doc.terms[i]["discount_percent"] == 0){
```
Comment (vjFaLk) [inline erpnext/accounts/doctype/payment_terms_template/payment_terms_template.js]: Change to `else if`

File: erpnext/accounts/doctype/payment_entry/payment_entry.py [hunk 7]
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
...
+		[{"reference_name": frm.doc.references[i].reference_name,
+		"reference_doctype": frm.doc.references[i].reference_doctype,
+		"posting_date": frm.doc.posting_date,
+		"company": frm.doc.company
+		},{... ]
+	"""
+	deductions = []
+	if isinstance(discount_double_check, basestring):
+		discount_double_check = json.loads(discount_double_check)
+	for i in discount_double_check:
```
Comment (vjFaLk) [inline erpnext/accounts/doctype/payment_entry/payment_entry.py]: `i` is a Bad variable name

**Comment by agritheory** [comment]
7/11/18 update: I've been refactoring this code into isolated chunks in the 10-line-size. Since this was a originally a custom script too much of the code base was in JS and not enough on the server. I'm working that out now.

**Comment by agritheory** [comment]
Test modified is in payment terms template to account for the additional columns relating to discount.

**Comment by agritheory** [comment]
Thought this was ready to go but there's a Travis error in the test that isn't showing up in my local version. Will update by EOD EDT 8/1/18.

**Comment by agritheory** [comment]
I've merged with the latest versions of Frappe and ERPnext locally and cannot replicate any of the conflicts. Running test as `bench run-tests --doctype "Payment Entry"` on a fresh site yields no errors.
@vjFaLk or @rmehta if you could resume code review and/or mark for merge I think this is ready.

**Comment by agritheory** [comment]
@vjFaLk Travis is now passing. The codacy requests don't seem worth another commit. Please advise.

**Comment by vjFaLk** [comment]
Yesus, this is a commit mess. Will need to do a lot of git wizardry. Lemme pull your branch and clean it up

**Comment by heatherpotter** [comment]
Please don't go stale! I really really need this!

**Comment by vjFaLk** [comment]
@nabinhait sir pls

**Comment by FinForceConsulting** [comment]
@agritheory, did you create a new PR for this? please share link

**Comment by gdlrsilv3** [comment]
hi @agritheory  , were you able to add the payment term discounts functionality to ERP Next?

**Pull request closed without merging**